shadowsocks/stream.go

```diff
@@ -68,16 +135,21 @@ type shadowsocksWriter struct {
 
 // NewShadowsocksWriter creates a Writer that encrypts the given Writer using
 // the shadowsocks protocol with the given shadowsocks cipher.
-func NewShadowsocksWriter(writer io.Writer, ssCipher shadowaead.Cipher) Writer {
-	return &shadowsocksWriter{writer: writer, ssCipher: ssCipher}
+func NewShadowsocksWriter(writer io.Writer, ssCipher shadowaead.Cipher) *Writer {
+	return &Writer{writer: writer, ssCipher: ssCipher, saltGenerator: randomSaltGenerator{ssCipher.SaltSize()}}
+}
+
+// SetSaltGenerator sets the salt generator to be used. Must be called before the first write.
+func (sw *Writer) SetSaltGenerator(saltGenerator SaltGenerator) {
+	sw.saltGenerator = saltGenerator
```

Comment:
With the new interface, there's nothing preventing you from passing a salt generator for the  wrong salt size. I prefer the previous interface where  you pass the slice to GetSalt, because then the caller of GetSalt is in control. It's harder to misuse.

It's also more flexible if the caller controls allocations. They could use a buffer pool, for example.

Reply:
OK, I've changed it to take a slice argument.